### DISCUSSION TITLE
Failed to bind to address: address already in use

### DISCUSSION THREAD

**Discussion by salhidev** [microsoft/aspire#1615] [Q&A] [answered]
I have a solution with multi microservices + frontend React SPA app, I choose to run the SPA + 1 microservice (authAPI) on aspire.

The solution uses some libraries, so I need to run as **project reference** locally. I can run on **debug** mode, but I'd need to generate a new version of the libraries using .Net8 which our devops agent is not configure yet to run on. But my assumption is this should work on proj ref as well.

When I run the AppHost I get an IOException : 
![image](https://github.com/dotnet/aspire/assets/17907269/02d6130c-acb3-466d-9c13-bcd63a136098)

Aspire loads fine + SPA, but Auth service is not running : 
![image](https://github.com/dotnet/aspire/assets/17907269/a6f33fee-0e5b-40a5-ac55-fb08cdc47c94)

Logs for authAPI : 
![image](https://github.com/dotnet/aspire/assets/17907269/44518375-7d12-4eb8-93cc-d0078dad84f7)

**Comment by davidfowl** [comment]
Can you provide more information? What does the apphost code looks like? What do the launch settings look like for the failing projects? I see authApi has no endpoints, which makes me think there's something strange going on with the launchSettings.json for that project.

**Comment by davidfowl** [comment]
It's very likely you're running into this https://github.com/dotnet/aspire/issues/225

**Comment by salhidev** [comment]
@davidfowl I also suspect it's launchSetting,json, I ad a new WeatherAPI and used WithServiceBinding() and it's working fine. My existing launchSettings.cs for authAPI is : 

 ```
{
  "iisSettings": {
    "windowsAuthentication": false,
    "anonymousAuthentication": true,
    "iisExpress": {
      "applicationUrl": "http://localhost:61060/Authentication",
      "sslPort": 0
    }
  },
  "$schema": "http://json.schemastore.org/launchsettings.json",
  "profiles": {
    "IIS Express": {
      "commandName": "IISExpress",
      "launchBrowser": true,
      "launchUrl": "swagger",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Local"
      }
    },
    "AuthenticationService": {
      "commandName": "Project",
      "launchBrowser": true,
      "launchUrl": "api/values",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Local"
      },
      "applicationUrl": "http://localhost:5000"
    }
  }
}
```

**Comment by davidfowl** [comment]
It's trying to use the "IIS Express" profile and there's no application URL. You can specify the launch profile in your app host code using `WithLaunchProfile("AuthenticationService")`. That'll choose the other URL.

I'm not sure that we support launching IIS express (we especially don't from the command line) @BillHiebert ?

**Comment by salhidev** [comment]
@davidfowl with iisexpress, the application url allow to specify the endpoint name : http://localhost:61060/Authentication
So swagger will launch under : http://localhost:61060/Authentication/swagger/index.html
http & https profiles don't allow this. Is there a way to include the endpoint name in the url of endpoints ?

**Comment by DamianEdwards** [comment]
Your screenshot seems to show you starting two hosts in a single application, one using the new `Host.CreateApplicationBuilder` and immediately again using the older `CreateHostBuilder` pattern. This will result in the second host failing as it attempts to bind to the port being used by the first. You should convert your project code to use the newer approach (`Host.CreateApplicationBuilder`) only.

**Comment by salhidev** [comment]
@DamianEdwards Thanks, done, I refactor to use WebApplication.CreateBuilder throughout.
![image](https://github.com/dotnet/aspire/assets/17907269/cf335047-856e-4495-bb47-e7e2421e9734)

But, I'm still struggling on getting the servicename on the url using launchSettings profiles : 
![image](https://github.com/dotnet/aspire/assets/17907269/5a235aed-8b30-4589-9d5b-fd9912e745ac)

The endpoint is showing as **http://localhost:5280/swagger** 
instead of : **http://localhost:5280/Authentication/swagger**
![image](https://github.com/dotnet/aspire/assets/17907269/a74fa18f-49a4-45a2-8d6e-66afa961154b)
 
Any insights will be much appreciated.

**Comment by DamianEdwards** [comment]
The `"launchUrl"` property isn't additive, it's either root-relative or absolute, so if you want it to launch to "/Authentication/swagger" you need to set it to that, i.e.:

```json
{
  "$schema": "https://json.schemastore.org/launchsettings.json",
  "profiles": {
    "http": {
      "commandName": "Project",
      "launchBrowser": true,
      "launchUrl": "Authentication/swagger",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Local"
      },
      "dotnetRunMessages": true,
      "applicationUrl": "http://localhost:5280/Authentication"
    },
    "https": {
      "commandName": "Project",
      "launchBrowser": true,
      "launchUrl": "Authentication/swagger",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Local"
      },
      "dotnetRunMessages": true,
      "applicationUrl": "https://localhost:7104/Authentication"
    }
  }
}
```

**Comment by salhidev** [comment]
@DamianEdwards This indeed changes the url : 
![image](https://github.com/dotnet/aspire/assets/17907269/3bd3d377-13b1-4233-8bcc-cced35e4b506)

But clicking on it doesn't load swagger : 
![image](https://github.com/dotnet/aspire/assets/17907269/a89cd6cf-b715-4b9a-9e86-c787ff1a2745)
It still somehow binds to http://localhost:5280/swagger/index.html  : 
![image](https://github.com/dotnet/aspire/assets/17907269/dfe4e722-db08-4070-8ab0-b5d6d71d85c2)

**Comment by DamianEdwards** [comment]
This might be an issue with Swashbuckle itself not properly supporting `PathBase`. Note that setting `"applicationUrl"` to include a path section is fairly atypical  so I wouldn't be surprised if there's extra configuration of Swashbuckle required to make it work properly.

**Comment by salhidev** [comment]
@DamianEdwards Thanks, it looks like it's a swagger registration problem. @davidfowl I see in the screenshot [here](https://github.com/dotnet/aspire/issues/1431 ), you have the servicename "Weather" in the url of the webapi endpoint, does it it load swagger when you click on the endpoint ? Would you share the code ?
![image](https://github.com/dotnet/aspire/assets/17907269/15b99499-2003-4032-86a7-dbe097bfd963)

**Reply by salhidev** [reply, marked as answer]
I was able to solve this using `app.UsePathBase("/Authentication");`
while making my swagger setup agnostic : 
```
app.UseSwagger();
app.UseSwaggerUI(options => {
    options.SwaggerEndpoint("v1/swagger.json", "My API V1");
});
`.WithLaunchProfile("Authentication");`
Authentication profile in launchSettings.json
```
  "Authentication": {
    "commandName": "Project",
    "launchBrowser": true,
    "launchUrl": "Authentication/swagger",
    "applicationUrl": "https://localhost:7104;http://localhost:5280",
    "environmentVariables": {
      "ASPNETCORE_ENVIRONMENT": "Local"
    }
  }
```

**Comment by paulomorgado** [comment]
What if I want to define a different port other than the one in the profile?

**Reply by DamianEdwards** [reply]
You can call `WithServiceBinding` on the resource builder (changing to `WithEndpoint` in future versions) and pass the port you want to use.

**Reply by paulomorgado** [reply]
Maybe I explained myself wrong. I want to not use the port in the profile and use another.

Using `WithServiceBinding(hostPort: 8888)` throws "Unrecognized scheme in server address 'tcp://localhost:63376'. Only 'http://' and 'https://' are supported."

Using `WithServiceBinding(hostPort: 8888, scheme: "http")` throws "One or more errors occurred. (Service binding annotation with name 'http' already exists.)"

Using `WithServiceBinding(hostPort: 8888, scheme: "http", name: "TEST")` does allow Aspire to start. But with the endpoint with the port in the profile (5281) and the endpoint with the port in the defined service binding.

And both are injected in service discovery:

http://_TEST.localhost:8888
http://_http.localhost:5281

No endpoint without the name prefix (how does service discovery use it?).

**Reply by DamianEdwards** [reply]
Ah you want to disable implicit endpoints from the launch profile then? Trying this now myself a few ways I can't actually get it to work. @davidfowl do you know if this is possible today? Even if I create an empty launch profile and do the following, the web app inherits the `ASPNETCORE_URLS` env var from the AppHost so fails to start.

```csharp
builder.AddProject<Projects.AspireApp6_Web>("webfrontend")
    .WithReference(apiService)
    .WithLaunchProfile("empty")
    .WithServiceBinding(hostPort: 8088, scheme: "http", env: "ASPNETCORE_URLS");
```

**Reply by paulomorgado** [reply]
I had tried something like that and it didn't work.

**Reply by jspaey** [reply]
FYI: I received the same address binding error when Kestrel endpoints were defined in the appsettings.json of the hosted project.
Removing the Kestrel section resolved the issue.

![image](https://github.com/user-attachments/assets/8b5a54ef-f4a0-41fd-89e4-eaef9ea992c7)

![image](https://github.com/user-attachments/assets/42290e41-c18f-46d0-8c14-e746c2586801)

![image](https://github.com/user-attachments/assets/a5a2d879-a4b7-4bef-b6c8-4f9abf9f6fdd)

![image](https://github.com/user-attachments/assets/ec15c1cd-a469-471f-8635-152564961719)
